pydis_site/apps/api/viewsets/bot/nomination.py

```diff
@@ -1,32 +1,224 @@
+
+    #### Request body
+    >>> {
+    ...     'reason': 'Updated reason for this nomination',
+    ...     'end_reason': 'Updated end_reason for this nomination',
+    ... }
+
+    Note: The request body may contain either or both fields.
+
+    See operation 1 for the response format and status codes.
+    """
     serializer_class = NominationSerializer
-    queryset = Nomination.objects.prefetch_related('author', 'user')
-    frozen_fields = ('author', 'inserted_at', 'user')
+    queryset = Nomination.objects.all()
+    filter_backends = (DjangoFilterBackend, SearchFilter, OrderingFilter)
+    filter_fields = ('user__id', 'actor__id', 'active')
+    frozen_fields = ('id', 'actor', 'inserted_at', 'user', 'ended_at')
+    frozen_on_create = ('ended_at', 'end_reason', 'active', 'inserted_at')
 
-    def update(self, request, *args, **kwargs):
+    def create(self, request, *args, **kwargs):
         """
-        DRF method for updating a Nomination.
+        DRF method for creating a Nomination.
 
         Called by the Django Rest Framework in response to the corresponding HTTP request.
         """
+        for field in request.data:
+            if field in self.frozen_on_create:
+                raise ValidationError({field: ['This field cannot be set at creation.']})
 
+        user_id = request.data.get("user")
+        if Nomination.objects.filter(active=True, user__id=user_id).exists():
+            raise ValidationError({'active': ['There can only be one active nomination.']})
+
+        serializer = self.get_serializer(
+            data=ChainMap(
+                request.data,
+                {"active": True}
+            )
+        )
+        serializer.is_valid(raise_exception=True)
+        self.perform_create(serializer)
+        headers = self.get_success_headers(serializer.data)
+        return Response(serializer.data, status=status.HTTP_201_CREATED, headers=headers)
+
+    def partial_update(self, request, *args, **kwargs):
+        """
+        DRF method for updating a Nomination.
+
+        Called by the Django Rest Framework in response to the corresponding HTTP request.
+        """
         for field in request.data:
             if field in self.frozen_fields:
                 raise ValidationError({field: ['This field cannot be updated.']})
 
         instance = self.get_object()
         serializer = self.get_serializer(instance, data=request.data, partial=True)
         serializer.is_valid(raise_exception=True)
+
+        data = serializer.validated_data
+
+        # There are three distinct PATCH scenarios we need to validate.
+        if instance.active and 'active' not in data:
+            # 1. We're updating an active nomination without ending it.
+            if request.data.get('end_reason'):
+                raise ValidationError(
+                    {'end_reason': ["An active nomination can't have an end reason."]}
+                )
+
+        elif instance.active and not data.get('active'):
```

Comment:
I think this makes the logic easier to read as it makes it clear that it's checking the value in the key rather than also potentially checking if `get()` returned `None`.

```suggestion
        elif instance.active and not data['active']:
```

Reply:
Yes, I'll change this. This is left-over from a previous version that was not an `elif`. That would mean that the existance of `'active'` was not guaranteed and a simple look-up could fail. The previous `if` makes that a logical impossibility, though.